QccPack-0.61-1/src/libQccPackAVQ/utils/man1/gyencode.1
.TH GYENCODE 1 "QCCPACK" ""
.SH NAME
gyencode \- encode using Gersho and Yano's AVQ algorithm
.SH SYNOPSIS
.B gyencode
.RB "[\|" \-i
.IR  adaption\_interval "\|]"
.RB "[\|" \-fc
.IR final\_codebook "\|]"
.I datfile
.I codebook
.I codebook\_coder
.I channelfile
.I sideinfofile
.SH OPTIONS
.TP
.BI \-i " adaption\_interval"
Integer.  Number of vectors in an adaption interval.  Default value = 1000.
.TP
.BI \-fc " final\_codebook"
String. Output the final state of the codebook after encoding to
.IR final\_codebook .
.SH DESCRIPTION
.LP
.B gyencode
implements the simple version of Gersho and Yano's algorithm.
The vectors of
.I datfile
(DAT format)
are encoded using the Gersho-Yano algorithm, with
.I codebook
(CBK format)
being the initial codebook.
.B gyencode
outputs a channel of VQ indices,
.I channelfile
(CHN format), and side information,
.I sideinfofile
(SID format).
.I codebook\_coder
(SQ format) gives the scalar quantizer used to produce the
side information (see below).
.LP
The Gersho-Yano algorithm operates as follows.
Over the vectors of each adaption interval (the length of which is
given by
.IR adaption\_interval),
the partition with
the greatest partial distortion is split. 
An iteration of the generalized Lloyd algorithm is run to
refine the locations of the two new codewords.  One codeword replaces
the split codeword; the other replaces the codeword with the smallest
partial distortion.
.LP
.B gydecode
is used for the corresponding decoding of 
.I channelfile 
and
.IR sideinfofile ;
.B avqrate
calculates the bit rate for the Gersho-Yano algorithm as represented by
this channel and the side information.
See
.BR gydecode (1)
and
.BR avqrate (1)
for more details.
.LP
Note: If the number of the vectors in
.I datfile
is not a multiple of 
.IR adaption\_interval ,
then 
the final adaption interval will contain fewer than
.I adaption\_interval 
vectors.
.LP
If option
.B \-fc
is specified, 
.B gyencode
outputs the final state of the codebook, taking into account all codebook
updates performed during coding, to
.IR final\_codebook .
.SH "SIDE INFORMATION"
.LP
The side information output by
.B gyencode
consists of a series of 
addresses and vectors.
.B gyencode
updates two vectors per adaption interval.
Thus, for each adaption interval,
.I sideinfofile
contains two vectors, which update the codebook,
as well as their corresponding addresses in the codebook.
.LP
The side-information file,
.IR sideinfofile ,
consists of a sequence of symbols of the following form:
.nf

.I "    <symbol type>" <symbol value>

.fi
where
.I <symbol type>
is 2 for an address, and 3 for an update vector.
The symbols are stored in pairs:  first an address which is then followed by
its correspomding update vector.
For an address,
.I <symbol type>
= 2, and
.I <symbol value>
is an integer (0 through codebook size - 1) denoting the address
in the codebook to which the following vector is assigned.
If
.I <symbol type>
= 3, then the current symbol is a vector.
In this case,
.I <symbol value>
= 
.IR "v[1] v[2] " \|.\|.\|. " v[dim]".
The scalar quantizer 
.I codebook\_coder 
(SQ format) is used to
quantize each component of the vector that updates the
codebook.
Thus, each
.I v[i]
is an index output from this scalar quantizer.
Note:
.I dim
is the vector dimension specified in the header of 
.I sideinfofile
(see
.BR QccPack (1)).
Note: in all instances,
.I <symbol value>
and 
.I <symbol type>
are stored in ASCII characters.
.LP
Note:
The value of
.I adaption\_interval
is stored in the header of
.I sideinfofile
as
.I N
(see
.BR QccPack (1)).
.IR sideinfofile .
.SH "SEE ALSO"
.BR gydecode (1),
.BR avqrate (1),
.BR QccPackAVQ (3),
.BR QccPack (3)

A. Gersho and M. Yano, "Adaptive Vector Quantization by Progressive
Codevector Replacement," Proc. ICASSP, pp. 133-136, 1985.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
A. Gersho and M. Yano, "Adaptive Vector Quantization by Progressive
Codevector Replacement," Proc. ICASSP, pp. 133-136, 1985.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackAVQ/utils/gyencode.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "gyencode.h"

#define AVQGY_DEFAULT_ADAPTION_INTERVAL  1000

#define USG_STRING "[-i %d:adaption_interval] [-fc %: %s:final_codebook] %s:datfile %s:codebook %s:codebook_coder %s:channelfile %s:sideinfofile"


int AdaptionInterval = AVQGY_DEFAULT_ADAPTION_INTERVAL;
QccDataset Datfile;
QccVQCodebook Codebook;
QccChannel Channel;
QccAVQSideInfo SideInfo;

int final_codebook_specified = 0;
QccString FinalCodebookName;



int main(int argc, char *argv[])
{
  int block_size;
  int current_block;
  int num_blocks;
  int vector_dimension;
  int final_block_size;
  

  QccInit(argc, argv);
  
  QccDatasetInitialize(&Datfile);
  QccVQCodebookInitialize(&Codebook);
  QccChannelInitialize(&Channel);
  QccAVQSideInfoInitialize(&SideInfo);
  
  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &AdaptionInterval, 
                         &final_codebook_specified,
                         FinalCodebookName,
                         Datfile.filename,
                         Codebook.filename,
                         SideInfo.codebook_coder.filename,
                         Channel.filename,
                         SideInfo.filename))
    QccErrorExit();
  
  if (AdaptionInterval < 1)
    {
      QccErrorAddMessage("%s: Adaption interval must be >= 1\n",
                         argv[0]);
      QccErrorExit();
    }
  
  Datfile.access_block_size = AdaptionInterval;
  if (QccDatasetStartRead(&Datfile))
    {
      QccErrorAddMessage("%s: Error calling QccDatasetStartRead",
                         argv[0]);
      QccErrorExit();
    }
  block_size = QccDatasetGetBlockSize(&Datfile);
  num_blocks = Datfile.num_vectors / block_size;
  final_block_size = Datfile.num_vectors - num_blocks*block_size;
  vector_dimension = Datfile.vector_dimension;
  
  if (QccVQCodebookRead(&Codebook))
    {
      QccErrorAddMessage("%s: Error calling QccVQCodebookRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (Codebook.codeword_dimension != vector_dimension)
    {
      QccErrorAddMessage("%s: Codebook %s and dataset %s have different vector dimensions\n",
                         Codebook.filename, Datfile.filename);
      QccErrorExit();
    }
  
  if (QccSQScalarQuantizerRead(&(SideInfo.codebook_coder)))
    {
      QccErrorAddMessage("%s: Error calling QccSQScalarQuantizerRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (SideInfo.codebook_coder.type != QCCSQSCALARQUANTIZER_GENERAL)
    {
      QccErrorAddMessage("%s: Only general scalar quantizer is supported for codebook coder",
                         argv[0]);
      QccErrorExit();
    }
  
  Channel.access_block_size = block_size;
  Channel.channel_length = Datfile.num_vectors;
  Channel.alphabet_size = 
    Codebook.num_codewords;
  if (QccChannelStartWrite(&Channel))
    {
      QccErrorAddMessage("%s: Error calling QccChannelStartWrite()",
                         argv[0]);
      QccErrorExit();
    }
  
  SideInfo.N = block_size;
  SideInfo.vector_dimension = vector_dimension;
  SideInfo.vector_component_alphabet_size =
    SideInfo.codebook_coder.num_levels;
  if (QccGetProgramName(SideInfo.program_name))
    {
      QccErrorAddMessage("%s: Error calling QccGetProgramName()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccAVQSideInfoStartWrite(&SideInfo))
    {
      QccErrorAddMessage("%s: Error calling QccAVQSideInfoStartWrite()",
                         argv[0]);
      QccErrorExit();
    }

  for (current_block = 0; current_block < num_blocks; current_block++)
    {
      if (QccDatasetReadBlock(&Datfile))
        {
          QccErrorAddMessage("%s: Error calling QccDatasetReadBlock()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccAVQGershoYanoEncode(&Datfile, &Codebook, &Channel, &SideInfo,
                                 block_size, NULL, NULL))
        {
          QccErrorAddMessage("%s: Error calling QccAVQGershoYanoEncode()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccChannelWriteBlock(&Channel))
        {
          QccErrorAddMessage("%s: Error calling QccChannelWriteBlock()",
                             argv[0]);
          QccErrorExit();
        }
    }
  
  /*  Process final block, which may have size < block_size  */
  if (final_block_size)
    {
      Datfile.access_block_size =
        Channel.access_block_size = 
        block_size = final_block_size;
      
      if (QccDatasetReadBlock(&Datfile))
        {
          QccErrorAddMessage("%s: Error calling QccDatasetReadBlock()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccAVQGershoYanoEncode(&Datfile, &Codebook, &Channel, &SideInfo,
                                 block_size, NULL, NULL))
        {
          QccErrorAddMessage("%s: Error calling QccAVQGershoYanoEncode()",
                             argv[0]);
          QccErrorExit();
        }
      if (QccChannelWriteBlock(&Channel))
        {
          QccErrorAddMessage("%s: Error calling QccChannelWriteBlock()",
                             argv[0]);
          QccErrorExit();
        }
    }
  
  
  if (QccAVQSideInfoEndWrite(&SideInfo))
    {
      QccErrorAddMessage("%s: Error calling QccAVQSideInfoEndWrite()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccChannelEndWrite(&Channel))
    {
      QccErrorAddMessage("%s: Error calling QccChannelEndWrite()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccDatasetEndRead(&Datfile))
    {
      QccErrorAddMessage("%s: Error calling QccDatasetEndRead()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (final_codebook_specified)
    {
      QccStringCopy(Codebook.filename, FinalCodebookName);
      if (QccVQCodebookWrite(&Codebook))
        {
          QccErrorAddMessage("%s: Error calling QccVQCodebookWrite()",
                             argv[0]);
          QccErrorExit();
        }
    }
  
  QccDatasetFree(&Datfile);
  QccVQCodebookFree(&Codebook);
  QccChannelFree(&Channel);

  QccExit;
}
